### ISSUE TITLE
images aren't serialized

### ISSUE THREAD

**Issue by PerBothner** [jerch/xterm-addon-image#47]
I got xterm.js + xterm-add-image working under [DomTerm](https://domterm.org/xtermjs.html). However, it doesn't appear a sixel-produced image is saved by the serialize addon.
 
Specifically, I printed some images (using lsix), and then did a "detach session". When I then did "attach" on the session, the images were gone.

Serialization is important not only for detach+attach, but also when dragging a window to a new top-level window - this also uses serialization on most platforms.

In "regular" DomTerm (not based on xterm.js), an image is represented as a `<img>` element with a `data:` URL. When a buffer is serialized as HTML, this automatically handles such images.  (FWIW, I'm actively working on improving the sixel handling in regular DomTerm.)

**Comment by jerch** [comment]
@PerBothner Drafted a quick demo PR to test the idea above (see #48). Ofc this cannot be used until the serialize addon sees major changes. Also the image re-encoding is very rough atm (using IIP for every single cell, which is not how it should be done).

**Comment by jerch** [comment]
There is prolly a simpler solution possible - the image addon currently always draw on top of FG/BG content not erasing it. Thus it should be possible to keep most of the serialize addon as it is, but invoke a full line image serializer for every line. This has several advantages:
- serialize addon needs only tiny change
- no complicated fuse logic needed
- no stitching artefacts in x-progression
- most of the image render logic can be re-used

Will see if that works better...

**Comment by jerch** [comment]
Addendum regarding a proper protocol:

It turns out that IIP has all the basics to guarantee a stable cell coverage, if dimensions are set with plain numbers denoting row/col extend, eg. `width=80;height=1` for a single terminal line of 80 cols. It still needs the overprinting patch of 64244d8 to level out floating rounding issues, in fact the floor/ceil logic must be applied all over the place to avoid stitching/striping artefacts. This might be some food for a better protocol, most likely we need to address into fractions of a cell, as already suggested by @egmontkob years ago.

Nonetheless IIP with col/row metrics works good enough atm to get a working solution for serialization running. :smile_cat:

**Comment by jerch** [comment]
The QOI image format might be a good choice for image serialization, as it provides very fast encoding/decoding.

Some early numbers testing QOI decoding speed with the sixel-bench test:
```
qoi: 2.72s (48.06 MB/s, 251.76 fps)
png: 7.56s (18.88 MB/s, 90.62 fps)
jpg: 2.25s (7.18 MB/s, 304.94 fps)
sixel: 3.03s (39.16 MB/s, 226.2 fps)
```
(Note: The JPEG numbers are flawed, as I did a simple 75% conversion, which reduces data from ~130MB to 17MB for jpeg.)

Did not test yet the encoding path, but I expect this to be much faster than any PNG/JPEG encoding.

**Comment by PerBothner** [comment]
A simple way to serialize a canvas is to use `canvas.toDataUrl()`. That seems simple, portable, and reasonably efficient.
To unserialize, you can create a new `Image`, set the `src` attribute, and optionally `drawImage` into the canvas. There may be a way to avoid creating a temporrary `Image` object.
You still have to decide on an escape sequence to wrap the data url in.

**Comment by jerch** [comment]
Haha - stumbled over the next hassle - fast bytearray to string conversion. I really dont get it why JS has no fast `Uint8Array.toByteString` method, thats literally a malloc + memcpy away for the engine. But nope, we have to use slow JS looping with `String.fromCharCode` calls, :see_no_evil: 
Well did not investigate much yet beside testing various concat tricks, thats what I ended up for now:
```typescript
const data = ... // uint8 array from encoder
let p = 0;
let s = '';
while (p < data.length) {
  s += String.fromCharCode.apply(null, data.subarray(p, p += 65536) as any);
}
```
Also `TextDecoder` seems to be on par with this, but I am not sure yet, whether it will run into issues for certain charcodes. Needs some more testing...

Some interim results:
- QOI encoding is 2.5 to 5 times faster than toDataUrl/PNG, depending on the image at hand. Thats a lesser gain than I was hoping for, but still remarkable.
- The implementation overhead with custom encoders is quite big, but it is only ~10kB bundled, with ~3kB in wasm.
- `getImageData` gets much faster with a software canvas (`willReadFrequently` setting), and is much more reliable and stable in timings than any webgl hacks. It is also not hardware dependent. Clearly the better choice.

**Comment by jerch** [comment]
@PerBothner Imho QOI is the right decision, as it is able to serialize things 2.5 - 5x faster than with browser methods (see full roundtrip numbers and explanation here: https://github.com/jerch/xterm-addon-image/pull/48#issuecomment-1484145335).

**Comment by jerch** [comment]
@PerBothner
Hmm, currently the serialize plugin is pretty slow, it takes ~170 ms to serialize a 10k scrollbuffer of 87 cols (tested with `ls -lR /usr/` generating ~500kB of string data). Thats ~8 times slower than a wasm version with almost the same functionality (takes ~20 ms).
Therefore I might ditch the idea to extend the serialize addon, instead a rewrite seems more appropriate to me. It makes it also easier to rethink the serialization problem and come up with a proper interface, where custom extension can hook into.

**Comment by PerBothner** [comment]
For grapheme clusters, why do you need to store codepoint attributes, assuming you're using either dom or canvas renderers? The browser takes care of that. You may need to know cluster boundaries but you can detect those at the same time you're detecting double-width characters, using an efficient trie lookup (which would probably be a good candidate for wasm).

_"If those ever will make it back into xterm.js"_ If I get to the point of switching DomTerm to use the xterm.js engine by default, I'm assuming it will have to be a fork, since the xterm.js maintainers don't seem to be interested in rich data or variable-width text.

**Comment by jerch** [comment]
> For grapheme clusters, why do you need to store codepoint attributes, assuming you're using either dom or canvas renderers? The browser takes care of that. You may need to know cluster boundaries but you can detect those at the same time you're detecting double-width characters, using an efficient trie lookup (which would probably be a good candidate for wasm).

Yes, I meant that lookup data for the unicode codepoint range to eval incoming chars. I dont quite remember, what I used back then, but it was quite big (I think it was a condensed bitmap for fast lookup). So no, I was not storing additional attributes in the buffer, it was just the segmentation logic, that needed additional lookup data.

Btw I think other TEs implementing grapheme support put the segmentation burden early right after the utf-8 decoding before doing `wcwidth`, thus they keep things cell aligned. Technically thats a `grapheme_wcwidth` in the end, and imho the only way to keep things within that nasty grid metaphor.
The impl impact is quite remarkable though - unicode does not limit graphemes in x direction, so it is theoretically possible to end up with cells >2 half widths. Thats what most TEs will have issues with, and so does xterm.js in its renderer logic. This gets really horrible for heavily stacking language systems with tons of subscripted perceivable characters (hebrew, arabic, hindi languages etc.), widths may even be in fractions of a cell here. But unicode did not bother with formalizing that at all, they left it on purpose to font developers to get the best visual output. Thats really a nightmare for proper grapheme dealing on a grid system, and I have not seen anyone to fully address this. So all TEs are currently broken in their unicode support.

Anyway, its kinda offtopic here.

**Comment by PerBothner** [comment]
_Off-topic - about grapheme clusters:_ The table for codepoint attributes used in DomTerm is a Uint32Array of length 12764, initialized from a 4032-character atob string. I consider that acceptable.

Basically DomTerm treats a cluster as wide if any of the constituents is wide, or it sees an emoji_presentation_selector or a regional_indicator pair.

> This gets really horrible for heavily stacking language systems with tons of subscripted perceivable characters

That is why terminals need to figure out how to handle variable-width fonts. On my roadmap is a wrapper for input editors (such as GNU readline) where the terminal uses a variable-width font (layout handled by the browser engine), but it pretends to the application that it's all regular fixed-width columns. This would be enabled by some extra shell integration flags without applications having to be otherwise modified. Applications that know they are running under DomTerm can request that a variable-width font be used for output.
